src/gluonts/dataset/repository/_m3.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import json
+import os
+import warnings
+from pathlib import Path
+from typing import NamedTuple
+
+import numpy as np
+import pandas as pd
+
+from gluonts.dataset.repository._util import metadata, save_to_file, to_dict
+from gluonts.gluonts_tqdm import tqdm
+
+
+def check_dataset(dataset_path: Path, length: int):
+    # check that things are correct
+    from gluonts.dataset.common import load_datasets
+
+    ds = load_datasets(
+        metadata=dataset_path,
+        train=dataset_path / "train",
+        test=dataset_path / "test",
+    )
+
+    assert ds.test is not None
+    assert len(list(ds.train)) == length
+    assert len(list(ds.test)) == length
+
+    assert ds.metadata.prediction_length is not None
+
+    for ts_train, ts_test in tqdm(
+        zip(ds.train, ds.test), total=length, desc="checking consistency"
+    ):
+        train_target = ts_train["target"]
+        test_target = ts_test["target"]
+        assert (
+            len(train_target)
+            == len(test_target) - ds.metadata.prediction_length
+        )
+        assert np.all(train_target == test_target[: len(train_target)])
+
+
+def generate_m3_dataset(dataset_path: Path, m3_freq: str):
+    from gluonts.dataset.repository.datasets import default_dataset_path
+
+    m3_xls_path = default_dataset_path / "M3C.xls"
+    if not os.path.exists(m3_xls_path):
+        raise RuntimeError(
+            f"The m3 data is available at https://forecasters.org/resources/time-series-data/m3-competition/ "
+            f"Please download the file and copy the files to this location: {m3_xls_path}"
+        )
```

Comment:
Couldn't the original data be downloaded automatically?

Reply:
This is what happens for M4 at least: https://github.com/awslabs/gluon-ts/blob/master/src/gluonts/dataset/repository/_m4.py

Reply:
The server is blocking automated requests. We can spoof the user agent, but that seems a bit too dodgy for including it in the package.

Reply:
Ideally, we would host the file ourselves, but there is no licence on the website. It is however also included in this gpl3 package: https://github.com/robjhyndman/Mcomp/tree/master/data